sources/api/schnauzer/src/v2/template.rs

```
@@ -0,0 +1,159 @@
+    pub fn extension_requirements(&self) -> impl Iterator<Item = ExtensionRequirement> + '_ {
+        self.required_extensions
+            .iter()
+            .map(|(extension_name, extension_requirements)| {
+                ExtensionRequirement::from_template_requirements(
+                    extension_name,
+                    extension_requirements,
+                )
+            })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, Hash)]
+pub struct ExtensionRequirement {
+    pub name: ExtensionName,
+    pub version: ExtensionVersion,
+    pub helpers: Vec<HelperName>,
+}
+
+impl ExtensionRequirement {
+    fn from_template_requirements(
+        extension_name: &ExtensionName,
+        requirements: &TemplateExtensionRequirements,
+    ) -> Self {
+        match requirements {
+            TemplateExtensionRequirements::Version(version) => ExtensionRequirement {
+                name: extension_name.clone(),
+                version: version.clone(),
+                helpers: vec![],
+            },
+            TemplateExtensionRequirements::VersionAndHelpers(extension_requirements) => {
+                ExtensionRequirement {
+                    name: extension_name.clone(),
+                    version: extension_requirements.version.clone(),
+                    helpers: extension_requirements.helpers.clone().unwrap_or_default(),
+                }
+            }
+        }
+    }
+}
+
+impl FromStr for Template {
+    type Err = error::Error;
+
+    fn from_str(input_str: &str) -> Result<Template> {
+        // Potential improvement here would be to use `ariadne` crate to emit more helpful error messages
+        let mut parsed_pairs = TemplateParser::parse(Rule::template, input_str)
+            .context(error::GrammarParseSnafu)?
+            .flatten();
+
+        let frontmatter_pair = parsed_pairs
+            .clone()
+            .find(|p| p.as_rule() == Rule::toml_document)
```

I know that `find()` short-circuits when it finds a match; is there a reason we would need to ensure no more matches exist?  That is: should there only be a single match?  Do we need to throw an error if there are more?